### PR TITLE
Add capability to compute counterfactuals per point

### PR DESCRIPTION
**Pull request by gaugup**
## Description

To be able to compute counterfactuals from RAI dashboard we need the backend code in `counterfactual_manager` to be able to generate counterfactuals and feature importance per point.  This PR adds that functionality and related unit tests.

## Checklist

- [ ] I have added screenshots above for all UI changes.
- [ ] I have added e2e tests for all UI changes.
- [ ] Documentation was updated if it was needed.

### CODE REVIEW

File: responsibleai/responsibleai/managers/counterfactual_manager.py [hunk 1]
```diff
@@ -577,6 +577,67 @@ def compute(self):
...
+    def request_counterfactuals(self, query_id: str, data: Any):
+        """Return the counterfactuals for a given point.
+
+        :param query_id: The query id for the finding the
+                         counterfactual config.
+        :type query_id: str
+        :param data: The data point for which the counterfactuals
+                     need to be generated.
+        :type data: Any
+        :return: The serialized version counterfactuals for the
```
Comment (imatiach-msft) [inline responsibleai/responsibleai/managers/counterfactual_manager.py:589]: this description is a bit confusing, shouldn't it be the serialized counterfactuals as a dictionary, and not a version, or did I misunderstand?

**Review by imatiach-msft — APPROVED** [review]

File: responsibleai/tests/counterfactual/test_counterfactual_advanced_features.py [hunk 2]
```diff
@@ -153,6 +154,87 @@ def test_counterfactual_manager_save_load(self, tmpdir):
...
+            test=X_test.iloc[0:10],
+            target_column='target',
+            task_type='classification')
+
+        rai_insights.counterfactual.add(
+            total_CFs=10, desired_class=2,
+            features_to_vary=[feature_names[0]],
+            permitted_range={feature_names[0]: [2.0, 5.0]},
+            feature_importance=feature_importance)
+        rai_insights.counterfactual.add(
```
Comment (imatiach-msft) [inline responsibleai/tests/counterfactual/test_counterfactual_advanced_features.py:179]: nit: maybe put this in a loop? it looks like same values are used except for desired class, so you can loop through desired_class = [1, 2] perhaps.

File: responsibleai/tests/counterfactual/test_counterfactual_advanced_features.py [hunk 3]
```diff
@@ -153,6 +154,87 @@ def test_counterfactual_manager_save_load(self, tmpdir):
...
+
+        assert counterfactual_obj_serialized['test_data'] is not None
+        assert counterfactual_obj_serialized['cfs_list'] is not None
+        if feature_importance:
+            assert counterfactual_obj_serialized[
+                'local_importance'] is not None
+        else:
+            assert counterfactual_obj_serialized[
+                'local_importance'] is None
+        assert counterfactual_obj_serialized['summary_importance'] is None
```
Comment (imatiach-msft) [inline responsibleai/tests/counterfactual/test_counterfactual_advanced_features.py:201]: nit: maybe you can loop through these as well - have a list of None and not None fields and assert them in a loop

File: responsibleai/responsibleai/managers/counterfactual_manager.py [hunk 4]
```diff
@@ -577,6 +577,67 @@ def compute(self):
                     cf_config.failure_reason = str(e)
                     raise e
 
+    def request_counterfactuals(self, query_id: str, data: Any):
+        """Return the counterfactuals for a given point.
+
+        :param query_id: The query id for the finding the
```
[the remark this answers is not in the document]
Comment (gaugup) [reply responsibleai/responsibleai/managers/counterfactual_manager.py:583]: Removed. Done.

File: responsibleai/responsibleai/managers/counterfactual_manager.py (hunk 1 above)
Comment (gaugup) [reply responsibleai/responsibleai/managers/counterfactual_manager.py:589]: Thanks for pointing this out. Changed to following:- 

        :return: An object of type CounterfactualData with
                 counterfactuals for the given data point.
        :rtype: CounterfactualData

File: responsibleai/tests/counterfactual/test_counterfactual_advanced_features.py (hunk 2 above)
Comment (gaugup) [reply responsibleai/tests/counterfactual/test_counterfactual_advanced_features.py:179]: I guess the configs aren't too many at this point. Maybe we can do the suggested way once we have more such configs.

File: responsibleai/responsibleai/managers/counterfactual_manager.py (hunk 1 above)
Comment (gaugup) [reply responsibleai/responsibleai/managers/counterfactual_manager.py]: I made the API return object of type CounterfactualData. Thanks for pointing this out.

File: responsibleai/tests/counterfactual/test_counterfactual_advanced_features.py (hunk 3 above)
Comment (gaugup) [reply responsibleai/tests/counterfactual/test_counterfactual_advanced_features.py]: Done.

**Pull request merged**